File: Week_03/linked_list_reverse_print/Solution.java
package linked_list_reverse_print;

import linked_list.ListNode;

public class Solution {

    private int[] result;

    public int[] reversePrint(ListNode head) {
        reverse(head, 0);
        return result;
    }

    private void reverse(ListNode head, int count) {
        if (head == null) {
            result = new int[count];
            return;
        }
        reverse(head.next, count + 1);
        result[result.length - count - 1] = head.val;
    }
}

File: Week_03/linked_list_reverse_print/Solution.py
# 递归到尾后，再把值从尾添加到数组
class Solution1:
    # [] + [] 涉及到数据搬移，效率低
    # def reversePrint(self, head: ListNode) -> List[int]:
    #     return self.reversePrint(head.next) + [head.val] if head else []

    def reversePrint(self, head: ListNode) -> List[int]:
        # 递归到尾后，再把值从尾添加到数组
        def reverse(head):
            if head:
                reverse(head.next)
                res.append(head.val)

        res = []
        reverse(head)
        return res


# 循环
class Solution2:
    def reversePrint(self, head: ListNode) -> List[int]:
        res = []
        while head:
            res.append(head.val)
            head = head.next
        return res[::-1]
